indexing-service/src/main/java/io/druid/indexing/DatasourcePathSpec.java

```diff
@@ -0,0 +1,104 @@
+import io.druid.indexing.common.task.NoopTask;
+import io.druid.indexing.common.task.Task;
+import io.druid.timeline.DataSegment;
+import io.druid.timeline.TimelineObjectHolder;
+import io.druid.timeline.VersionedIntervalTimeline;
+import io.druid.timeline.partition.PartitionChunk;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.mapreduce.Job;
+import org.apache.hadoop.mapreduce.lib.input.MultipleInputs;
+
+import java.io.IOException;
+import java.util.List;
+
+public class DatasourcePathSpec implements PathSpec
+{
+  private TaskActionClientFactory taskActionClientFactory;
+  private SegmentLoadSpec spec;
+  private ObjectMapper mapper;
+
+  public DatasourcePathSpec(
+      @JacksonInject TaskActionClientFactory taskActionClientFactory,
+      @JacksonInject ObjectMapper mapper,
+      @JsonProperty("ingestSpec")SegmentLoadSpec spec
+  )
+  {
+    this.taskActionClientFactory = Preconditions.checkNotNull(taskActionClientFactory, "null TaskActionClientFactory");
+    this.mapper = Preconditions.checkNotNull(mapper, "null mapper");
+    this.spec = Preconditions.checkNotNull(spec, "null ingestSpec");
+  }
+
+  @Override
+  public Job addInputPaths(
+      HadoopDruidIndexerConfig config, Job job
+  ) throws IOException
+  {
+    Task task = new NoopTask("reingest", 0, 0, null, null);
+    TaskActionClient tac = taskActionClientFactory.create(task);
+    final List<DataSegment> usedSegments = tac.submit(new SegmentListUsedAction(spec.getDataSource(), spec.getInterval()));
+
+    VersionedIntervalTimeline<String, DataSegment> timeline = new VersionedIntervalTimeline<String, DataSegment>(
+        Ordering.<String>natural().nullsFirst()
+    );
+
+    for (DataSegment segment : usedSegments) {
+      timeline.add(segment.getInterval(), segment.getVersion(), segment.getShardSpec().createChunk(segment));
+    }
+    final List<TimelineObjectHolder<String, DataSegment>> timeLineSegments = timeline.lookup(
+        spec.getInterval()
+    );
+
+    final List<String> paths = Lists.newArrayList(
+        Iterables.concat(
+            Iterables.transform(
+                timeLineSegments,
+                new Function<TimelineObjectHolder<String, DataSegment>, Iterable<String>>()
+                {
+                  @Override
+                  public Iterable<String> apply(
+                      TimelineObjectHolder<String, DataSegment> input
+                  )
+                  {
+                    return Iterables.transform(
+                        input.getObject(),
+                        new Function<PartitionChunk<DataSegment>, String>()
+                        {
+                          @Override
+                          public String apply(PartitionChunk<DataSegment> input)
+                          {
+                            final DataSegment segment = input.getObject();
+                            return (String) segment.getLoadSpec().get("path");
```

Comment:
as noted earlier, I will put all of loadSpec. that should provide enough information to load the segment.